core/embed/trezorhal/board_capabilities.c

```diff
@@ -0,0 +1,81 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "board_capabilities.h"
+#include <string.h>
+#include "common.h"
+
+#define handle_fault(msg) \
+  (__fatal_error("Fault detected", msg, __FILE__, __LINE__, __func__))
+
+static uint8_t board_name[MODEL_NAME_MAX_LENGTH + 1] = {0};
+
+static struct BoardloaderVersion boardloader_version;
+
+const uint8_t *get_board_name() { return board_name; }
+
+const struct BoardloaderVersion *get_boardloader_version() {
+  return &boardloader_version;
+}
+
+void parse_boardloader_capabilities() {
+  uint8_t *pos = (uint8_t *)BOARD_CAPABILITIES_ADDR;
+  uint8_t *end = (uint8_t *)(BOARD_CAPABILITIES_ADDR + BOARD_CAPABILITIES_SIZE);
+
+  if (memcmp(pos, CAPABILITIES_HEADER, 4) != 0) {
+    return;
+  }
+
+  pos += 4;
+
+  // -2 for possible tag without any data
+  while (pos <= end - 2) {
```

Comment:
Is the `-2` really necessary?
It should work correctly with the following, no?
And we could kill the terminator tag completely.

```suggestion
  while (pos < end) {
```

Reply:
With `(pos < end)` you could read out-of-bound one byte behind the sector if `(pos == end-1)`.

Though we will use well-formed structure I thought then it's better be able to deal with malformed input to avoid any bugs.

The terminator is needed since the other two options are:

 * make and "empty tag" to have length until the end, which is kind of the same
 * rely that empty flash bytes are 0xFF behind the structure, so that it would be coincidentally seen as unknown tag of length 255 and cycle will end

It seems cleaner to have the terminator than the above two other options.

Reply:
Okay 👍